File: Data Structures & Algorithms/maximum-subarray/submission-1.cpp
class Solution{
public:
    int maxSubArray(vector<int>& nums){
        int currentSum=0;
        int best=INT_MIN;
        for(int i=0;i<nums.size();i++){
            currentSum+=nums[i];
            best=max(best,currentSum);
            if(currentSum<0){
                currentSum=0;
            }
        }
        return best;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.java
class Solution{
    public int maxProduct(int[] nums){
        int minProd=nums[0];
        int maxProd=nums[0];
        int best=nums[0];
        for(int i=1;i<nums.length;i++){
            int prevMin=minProd;
            int prevMax=maxProd;
            minProd=Math.min(nums[i],Math.min(prevMin*nums[i],prevMax*nums[i]));
            maxProd=Math.max(nums[i],Math.max(prevMin*nums[i],prevMax*nums[i]));
            best=Math.max(best,maxProd);
        }
        return best;
    }
}